shm_wrapper/shm_wrapper.c

```diff
@@ -511,159 +533,338 @@ void device_disconnect (int dev_ix)
+	}
+	
+	//release gp_sem
+	my_sem_post(gp_sem, "gamepad_mutex");
+	
+	return 0;
+}
+
+/*
+ * This function writes the given state of the gamepad to shared memory.
+ * Blocks on both the gamepad semaphore and device description semaphore (to check if gamepad connected).
+ * Arguments:
+ *    - uint32_t pressed_buttons: a 32-bit bitmap that corresponds to which buttons are currently pressed
+            (only the first 17 bits used, since there are 17 buttons)
+ *    - float *joystick_vals: array of 4 floats that contain the values to write to the joystick
+ * Returns 0 on success, -1 on failuire (if gamepad is not connected)
+ */
+int gamepad_write (uint32_t pressed_buttons, float *joystick_vals)
+{
+	//wait on rd_sem
+	my_sem_wait(rd_sem, "robot_desc_mutex");
+	
+	//if no gamepad connected, then release rd_sem and return
+	if (rd_shm_ptr->fields[GAMEPAD] == DISCONNECTED) {
+		log_printf(ERROR, "tried to write, but no gamepad connected");
+		my_sem_post(rd_sem, "robot_desc_mutex");
+		return -1;
+	}
+	
+	//release rd_sem
+	my_sem_post(rd_sem, "robot_desc_mutex");
+	
+	//wait on gp_sem
+	my_sem_wait(gp_sem, "gamepad_mutex");
+	
+	gp_shm_ptr->buttons = pressed_buttons;
+	for (int i = 0; i < 4; i++) {
+		gp_shm_ptr->joysticks[i] = joystick_vals[i];
+	}
+	
+	//release gp_sem
+	my_sem_post(gp_sem, "gamepad_mutex");
+	
+	return 0;
 }
 
 /*
-Should be called from all processes that want to know current state of the param bitmap (i.e. device handler)
-Blocks on the param bitmap semaphore for obvious reasons
-	- bitmap: pointer to array of 17 32-bit integers to copy the bitmap into
-No return value.
-*/
-void get_param_bitmap (uint32_t *bitmap)
+ * Should be called from all processes that want to know current state of the command bitmap (i.e. device handler)
+ * Blocks on the param bitmap semaphore for obvious reasons
+ * Arguments:
+ *    - uint32_t *bitmap: pointer to array of 17 32-bit integers to copy the bitmap into
+ * No return value.
+ */
+void get_cmd_map (uint32_t *bitmap)
 {
-	//wait on pmap_sem
-	my_sem_wait(pmap_sem, "pmap_sem");
+	//wait on cmd_map_sem
+	my_sem_wait(cmd_map_sem, "cmd_map_sem");
 	
 	for (int i = 0; i < MAX_DEVICES + 1; i++) {
-		bitmap[i] = shm_ptr->pmap[i];
+		bitmap[i] = dev_shm_ptr->cmd_map[i];
 	}
 	
-	//release pmap_sem
-	my_sem_post(pmap_sem, "pmap_sem");
+	//release cmd_map_sem
+	my_sem_post(cmd_map_sem, "cmd_map_sem");
 }
 
 /*
-Should be called from all processes that want to know device identifiers of all currently connected devices
-Blocks on catalog semaphore for obvious reasons
-	- dev_ids: pointer to array of dev_id_t's to copy the information into
-No return value.
-*/
+ * Should be called from all processes that want to know device identifiers of all currently connected devices
+ * Blocks on catalog semaphore for obvious reasons
+ * Arguments:
+ *    - dev_id_t *dev_ids: pointer to array of dev_id_t's to copy the information into
+ * No return value.
+ */
 void get_device_identifiers (dev_id_t *dev_ids)
```

Comment:
`void get_device_identifiers (dev_id_t dev_ids[MAX_DEVICES])`
(As declared in header file for consistency)

Reply:
👌